components/raftstore/src/store/peer.rs

```diff
@@ -1898,6 +1907,15 @@ where
                 if let LeaseState::Valid = state {
                     // If current peer has valid lease, then we could handle the
                     // request directly, rather than send a heartbeat to check quorum.
+
+                    let val = self.read_indx_resp.get(&m.from);
+                    if val.is_some() && *val.unwrap() == start_ts {
+                        // this message can be coalesced with the last one
+                        ctx.raft_metrics.read_indx_coal.inc();
+                    } else if self.next_proposal_index() > index {
```

Comment:
The purpose of this check is to determine whether there is a lock on any key within the region.  If there are no memory locks on this region, we can combine requests with the same timestamp. 

If the proposal index is greater than the commit index, it indicates that there are no pending writes in progress on this region and we can safely assume that there are no memory locks in this region.

Reply:
> If the proposal index is greater than the commit index, it indicates that there are no pending writes in progress on this region.

The memory lock would be checked at L1887 `on_step_read_index`, the `locked` field would be set if there are conflict memory locks. What's the meaning of the `pending writes` here? The `proposal index` of one raft group(or region) is not related to memory locks I think, a memory lock would released after the corresponding write command is applied.

Perhaps you could write a test case to verify if it works as expected.

Reply:
This looks out of the topic about adding metrics as said in the pull request's title. Please consider splitting them to another pull request or changing the title. By the way, our convention of PR titles is starting with "<module name>: ", which is not followed by this PR.

Reply:
>>My initial thought for resuing the checking results is:
We are on the same page regarding how to dedupe or coalesce the requests. However, check_ts_conflict_si or lock.rs doesn't have any information about region ID. We can add it later if we decide to go with this approach. 

I want this change to be very light weight and will not be checked in to master.  i am going to check if applied index is less than the proposal index to see if there are any pending memory locks.

Reply:
The `self.next_proposal_index() > self.get_store().applied_index()` is not related to memory lock check I think, actually the condition `proposal_index >= commit_index >= applied_index` holds most of the time.

It could not be told if a read request with `start_ts` and range `[start_key, end_key)` would encounter conflict memory locks by checking the raft indexes. It should be checked by the concurrency manager using `read_key_check` or `read_range_check` methods. A write record with smaller `commit_ts` could also be applied from a larger applied index.

To understand the whole process, it's better to write a unit test case to see if the result is expected.

Besieds, this PR could be merged to some test branches if it is used to do some pre-tests or demo. After the whole optimization is decided and ready, a tracking issue could be created and the related PRs including this metric one could be merged to the master branch.

Reply:
I had a chat with Rui on slack. I think we both agree that this metric will be helpful. In our set up, write qps is very low (~5k) and coprocessor read qps is (~2mil). I think that most of the time we are reading on the regions which are not updated too often.